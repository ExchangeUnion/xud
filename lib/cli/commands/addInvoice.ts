import { callback, loadXudClient } from '../command';
import { Arguments } from 'yargs';
import { AddInvoiceRequest } from '../../proto/xudrpc_pb';

export const command = 'addinvoice <value> [memo]';

export const describe = 'get orders from the order book';

export const builder = {
  value: {
    type: 'number',
  },
  memo: {
    type: 'string',
  },
};

export const handler = (argv: Arguments) => {
  const request = new AddInvoiceRequest();
  request.setValue(argv.value);
  request.setMemo(argv.memo);
  loadXudClient(argv).addInvoice(request, callback);
};
