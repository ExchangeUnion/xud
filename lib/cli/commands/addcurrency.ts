import { Arguments } from 'yargs';
import { callback, loadXudClient } from '../command';
import { AddCurrencyRequest } from '../../proto/xudrpc_pb';
import { SwapClients } from '../../types/enums';

export const command = 'addcurrency <currency> <swap_client> [decimal_places] [token_address]';

export const describe = 'add a currency';

export const builder = {
  currency: {
    description: 'The ticker symbol for the currency',
    type: 'string',
  },
  swap_client: {
    description: 'The payment channel network client for swaps',
    type: 'string',
    choices: ['LND', 'RAIDEN'],
  },
  decimal_places: {
    description: 'The places to the right of the decimal point of the smallest subunit (e.g. satoshi)',
    default: 8,
    type: 'number',
  },
  token_address: {
    description: 'The contract address for layered tokens such as ERC20',
    type: 'string',
  },
};

export const handler = (argv: Arguments) => {
  const request = new AddCurrencyRequest();
  request.setCurrency(argv.currency.toUpperCase());
  request.setSwapClient(Number(SwapClients[argv.swap_client]));
  request.setTokenAddress(argv.token_address);
  request.setDecimalPlaces(argv.decimal_places);
  loadXudClient(argv).addCurrency(request, callback);
};
