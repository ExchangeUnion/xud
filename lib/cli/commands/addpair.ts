import { Arguments } from 'yargs';
import { callback, loadXudClient } from '../command';
import { AddPairRequest } from '../../proto/xudrpc_pb';

export const command = 'addpair <base_currency> <quote_currency>';

export const describe = 'add a trading pair';

export const builder = {
  base_currency: {
    description: 'The currency bought and sold for this trading pair',
    type: 'string',
  },
  quote_currency: {
    description: 'The currency used to quote a price',
    type: 'string',
  },
};

export const handler = (argv: Arguments) => {
  const request = new AddPairRequest();
  request.setBaseCurrency(argv.base_currency.toUpperCase());
  request.setQuoteCurrency(argv.quote_currency.toUpperCase());
  loadXudClient(argv).addPair(request, callback);
};
