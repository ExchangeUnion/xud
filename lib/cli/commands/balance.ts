import { callback, loadXudClient } from '../command';
import { Arguments } from 'yargs';
import { BalanceRequest, BalanceResponse } from '../../proto/xudrpc_pb';
import { satsToCoinsStr } from '../utils';
import Table, { HorizontalTable } from 'cli-table3';
import colors from 'colors/safe';

const HEADERS = [
  colors.blue('Ticker'),
  colors.blue('Balance'),
  colors.blue('Pending Balance'),
];

const formatBalances = (balances: BalanceResponse.AsObject) => {
  const formatted: any[] = [];
  balances.balancesMap.forEach((balance) => {
    const element = [];
    element.push(balance[0], `${satsToCoinsStr(balance[1].balance)}`, `${satsToCoinsStr(balance[1].pendingOpenBalance)}`);
    formatted.push(element);
  });
  return formatted;
};

const createTable = () => {
  const table = new Table({
    head: HEADERS,
  }) as HorizontalTable;
  return table;
};

const displayBalances = (balances: BalanceResponse.AsObject) => {
  const table = createTable();
  const formatted = formatBalances(balances);
  formatted.forEach(balance => table.push(balance));
  console.log(colors.underline(colors.bold('\nBalance:')));
  console.log(table.toString());
};

export const command = 'balance [currency]';

export const describe = 'get total balance for a given currency';

export const builder = {
  currency: {
    describe: 'the currency to query for',
    type: 'string',
  },
};

export const handler = (argv: Arguments) => {
  const request = new BalanceRequest();
  if (argv.currency) {
    request.setCurrency(argv.currency.toUpperCase());
  }
  loadXudClient(argv).balance(request, callback(argv, displayBalances));
};
