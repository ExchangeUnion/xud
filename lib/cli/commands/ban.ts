import { Arguments, Argv } from 'yargs';
import { BanRequest } from '../../proto/xudrpc_pb';
import { callback, loadXudClient } from '../command';

export const command = 'ban <node_identifier>';

export const describe = 'ban a remote node';

export const builder = (argv: Argv) =>
  argv
    .positional('node_identifier', {
      description: 'the node key or alias of the remote node to ban',
      type: 'string',
    })
    .example('$0 ban 028599d05b18c0c3f8028915a17d603416f7276c822b6b2d20e71a3502bd0f9e0b', 'ban by node key')
    .example('$0 ban CheeseMonkey', 'ban by alias');

export const handler = async (argv: Arguments<any>) => {
  const request = new BanRequest();
  request.setNodeIdentifier(argv.node_identifier);
  (await loadXudClient(argv)).ban(request, callback(argv));
};
