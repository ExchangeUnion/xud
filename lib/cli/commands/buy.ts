import { Arguments, Argv } from 'yargs';
import { OrderSide } from '../../proto/xudrpc_pb';
import { placeOrderBuilder, placeOrderHandler } from '../placeorder';

export const command = 'buy <quantity> <pair_id> <price> [order_id]';

export const describe = 'place a buy order';

export const builder = (argv: Argv) => placeOrderBuilder(argv, OrderSide.BUY);

export const handler = async (argv: Arguments) => placeOrderHandler(argv, OrderSide.BUY);
