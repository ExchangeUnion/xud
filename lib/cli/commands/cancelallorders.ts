import { Arguments, Argv } from 'yargs';
import { CancelAllOrdersRequest, CancelAllOrdersResponse } from '../../proto/xudrpc_pb';
import { callback, loadXudClient } from '../command';

export const command = 'cancelallorders';

export const describe = 'removes all orders';

export const builder = (argv: Argv) => argv
    .example('$0 cancelallorders', describe);

const formatOutput = (response: CancelAllOrdersResponse.AsObject) => {
  if (response.removedOrderIdsList.length) {
    response.removedOrderIdsList.forEach((removedOrder => console.log(`Cancelled order with id ${removedOrder}`)));
  } else {
    console.log('No orders found');
  }
};

export const handler = async (argv: Arguments<any>) => {
  (await loadXudClient(argv)).cancelAllOrders(new CancelAllOrdersRequest(), callback(argv, formatOutput));
};
