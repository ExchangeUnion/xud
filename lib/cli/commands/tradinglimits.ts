import Table, { HorizontalTable } from 'cli-table3';
import colors from 'colors/safe';
import { Arguments, Argv } from 'yargs';
import { TradingLimitsRequest, TradingLimitsResponse } from '../../proto/xudrpc_pb';
import { callback, loadXudClient } from '../command';
import { satsToCoinsStr } from '../utils';

const HEADERS = [
  colors.blue('Currency'),
  colors.blue('Max Buy'),
  colors.blue('Max Sell'),
  colors.blue('Reserved Buy'),
  colors.blue('Reserved Sell'),
];

const formatTradingLimits = (tradingLimits: TradingLimitsResponse.AsObject) => {
  const formatted: any[] = [];
  tradingLimits.limitsMap.forEach((tradingLimitElement) => {
    const currency = tradingLimitElement[0];
    const tradingLimit = tradingLimitElement[1];
    const row = [];
    row.push(
      currency,
      `${satsToCoinsStr(tradingLimit.maxBuy)}`,
      `${satsToCoinsStr(tradingLimit.maxSell)}`,
      `${satsToCoinsStr(tradingLimit.reservedBuy)}`,
      `${satsToCoinsStr(tradingLimit.reservedSell)}`,
    );
    formatted.push(row);
  });
  return formatted;
};

const createTable = () => {
  const table = new Table({ head: HEADERS }) as HorizontalTable;
  return table;
};

export const displayLimits = (limits: TradingLimitsResponse.AsObject) => {
  const table = createTable();
  const formatted = formatTradingLimits(limits);
  formatted.forEach((formattedLimit) => table.push(formattedLimit));
  console.log(colors.underline(colors.bold('\nTrading Limits:')));
  console.log(table.toString());
};

export const command = 'tradinglimits [currency]';

export const describe = 'trading limits for a given currency';

export const builder = (argv: Argv) =>
  argv
    .option('currency', {
      describe: 'the currency to query for',
      type: 'string',
    })
    .example('$0 tradinglimits', 'get the trading limits for all currencies')
    .example('$0 tradinglimits BTC', 'get the trading limits for BTC');

export const handler = async (argv: Arguments<any>) => {
  const request = new TradingLimitsRequest();
  if (argv.currency) {
    request.setCurrency(argv.currency.toUpperCase());
  }

  (await loadXudClient(argv)).tradingLimits(request, callback(argv, displayLimits));
};
