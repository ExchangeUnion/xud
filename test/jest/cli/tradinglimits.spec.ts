import { displayLimits } from '../../../lib/cli/commands/tradinglimits';

jest.mock('colors/safe', () => {
  return {
    blue: (str: string) => str,
    underline: (str: string) => str,
    bold: (str: string) => str,
  };
});

describe('displayLimits', () => {
  const mockLog = jest.fn();
  console.log = mockLog;

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should print a table', () => {
    displayLimits({
      limitsMap: [
        [
          'BTC',
          {
            maxBuy: 12345,
            maxSell: 67890,
            reservedBuy: 25000,
            reservedSell: 75000,
          },
        ],
      ],
    });

    expect(mockLog.mock.calls).toMatchSnapshot();
  });
});
