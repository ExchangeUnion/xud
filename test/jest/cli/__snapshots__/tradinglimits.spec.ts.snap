// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displayLimits should print a table 1`] = `
Array [
  Array [
    "
Trading Limits:",
  ],
  Array [
    "┌──────────┬────────────┬───────────┬──────────────┬───────────────┐
│ Currency │ Max Buy    │ Max Sell  │ Reserved Buy │ Reserved Sell │
├──────────┼────────────┼───────────┼──────────────┼───────────────┤
│ BTC      │ 0.00012345 │ 0.0006789 │ 0.00025      │ 0.00075       │
└──────────┴────────────┴───────────┴──────────────┴───────────────┘",
  ],
]
`;
